miden/tests/integration/stdlib/math/poly512.rs

```diff
@@ -1,11359 +1,361 @@
+        // fill script for polynomial 1
+        let _ = writeln!(
+            polynomial_1_script,
+            "push.{}.{}.{}.{}",
+            polynomial_1[4 * i + 3],
+            polynomial_1[4 * i + 2],
+            polynomial_1[4 * i + 1],
+            polynomial_1[4 * i]
+        );
+        let _ = writeln!(polynomial_1_script, "loc_storew.{}", i);
+        polynomial_1_script.push_str("dropw\n");
+
+        // fill script for polynomial 2
+        let _ = writeln!(
+            polynomial_2_script,
+            "push.{}.{}.{}.{}",
+            polynomial_2[4 * i + 3],
+            polynomial_2[4 * i + 2],
+            polynomial_2[4 * i + 1],
+            polynomial_2[4 * i]
+        );
+        let _ = writeln!(polynomial_2_script, "loc_storew.{}", i + 128);
+        polynomial_2_script.push_str("dropw\n");
+
+        // fill script for checking the result
+        check_result_script.push_str("push.0.0.0.0\n");
+        let _ = writeln!(check_result_script, "loc_loadw.{}", i + 256);
+        let _ = writeln!(check_result_script, "push.{}", result_polynomial[4 * i]);
+        check_result_script.push_str("assert_eq\n");
+        let _ = writeln!(check_result_script, "push.{}", result_polynomial[4 * i + 1]);
+        check_result_script.push_str("assert_eq\n");
+        let _ = writeln!(check_result_script, "push.{}", result_polynomial[4 * i + 2]);
+        check_result_script.push_str("assert_eq\n");
+        let _ = writeln!(check_result_script, "push.{}", result_polynomial[4 * i + 3]);
+        check_result_script.push_str("assert_eq\n");
+    }
+
+    let script = format!(
+        "
+        use.std::math::poly512
+
+        proc.wrapper.384
+            {}
+
+            {}
+
+            locaddr.256 # output
+            locaddr.128 # input 1
+            locaddr.0 # input 0
+
+            exec.poly512::sub_zq
+
+            {}
+        end
+
+        begin
+            exec.wrapper
+        end
+    ",
+        polynomial_1_script, polynomial_2_script, check_result_script
+    );
+    script
+}
 
-    begin
-        exec.wrapper
-    end
-    ";
+#[test]
+fn test_poly512_mul_zq() {
+    let source = generate_test_script_mul_zq();
 
     let test = build_test!(source, &[]);
     test.get_last_stack_state();
+    // test.expect_stack(&[1337, 69]);
+}
+
+fn generate_test_script_mul_zq() -> String {
+    const POLYNOMIAL_LENGTH: usize = 512;
+    const WORDS: usize = 128;
+    const Q: u64 = 12289; // Prime Number
+
+    let polynomial_1 =
+        rand_utils::rand_array::<u64, POLYNOMIAL_LENGTH>().map(|v| BaseElement::new(v % Q));
+    let polynomial_2 =
+        rand_utils::rand_array::<u64, POLYNOMIAL_LENGTH>().map(|v| BaseElement::new(v % Q));
+
+    let result_polynomial: Vec<u64> = mul(&polynomial_1, &polynomial_2)
+        .iter()
+        .map(|v| v.as_int() % Q)
+        .collect();
+
+    let lower = result_polynomial[..512].to_vec();
+    let mut upper = result_polynomial[512..].to_vec();
+    upper.push(0);
```

Comment:
Might be easier to use [split_at](https://doc.rust-lang.org/std/primitive.slice.html#method.split_at) here.